### PR TITLE
Non-template propagator, Implement Global Propagator, no-op propagator

### LINKED ISSUE #540: Implement global propagators
As per specs : https://github.com/open-telemetry/opentelemetry-specification/blob/11cc73939a32e3a2e6f11bdeab843c61cf8594e9/specification/context/api-propagators.md

**Global Propagators**
The OpenTelemetry API MUST provide a way to obtain a propagator for each supported Propagator type. Instrumentation libraries SHOULD call propagators to extract and inject the context on all remote calls. Propagators, depending on the language, MAY be set up using various dependency injection techniques or available as global accessors.

### PR DESCRIPTION
**Pull request by lalitb**
Fixes #540 

## Changes

Please provide a brief description of the changes here.
 - In continuation with discussion in #644, remove the templatized propagator class,  so that the propagator is not compiled against its carrier/target type, and the carrier/target type is instead passed as `TextMapCarrier` object in Inject() and Extract() functions. This further allows not to create global/composite propagator per carrier/target type.  As an example, a global propagator once set, can be used to extract the context from (say) gRPC response, and pass the context to (say) HTTP request. 
 
 - Provide an implementation of global and No-op propagator.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [x] Changes in public API reviewed

### CODE REVIEW

File: api/include/opentelemetry/trace/propagation/global_propagator.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#pragma once
```
Comment (ThomsonTan) [inline api/include/opentelemetry/trace/propagation/global_propagator.h:1]: License header?

File: api/include/opentelemetry/trace/propagation/noop_propagator.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+{
+
+/**
+ * No-op implementation TextMapPropagator
+ */
+class NoOpPropagator : public TextMapPropagator
+{
+public:
+  /** Noop extract function does nothing and returns the input context */
+  virtual context::Context Extract(const TextMapCarrier & /*carrier*/,
```
Comment (ThomsonTan) [inline api/include/opentelemetry/trace/propagation/noop_propagator.h:19]: `virtual` keyword is unnecessary here?

File: api/include/opentelemetry/trace/propagation/text_map_propagator.h [hunk 3]
```diff
@@ -11,32 +11,33 @@ namespace trace
 namespace propagation
 {
 
+// TextMapCarrier is the storage medium used by TextMapPropagator.
+class TextMapCarrier
+{
+
```
Comment (ThomsonTan) [inline api/include/opentelemetry/trace/propagation/text_map_propagator.h:17]: nit: remove empty line.

File: api/include/opentelemetry/trace/propagation/text_map_propagator.h [hunk 4]
```diff
@@ -11,32 +11,33 @@ namespace trace
...
+public:
+  /*returns the value associated with the passed key.*/
+  virtual nostd::string_view Get(const nostd::string_view &key) const noexcept = 0;
```
Comment (ThomsonTan) [inline api/include/opentelemetry/trace/propagation/text_map_propagator.h:20]: Pass `nostd::string_view` by value?

File: api/test/trace/propagation/http_text_format_test.cc [hunk 5]
```diff
@@ -137,11 +143,55 @@ TEST(TextMapPropagatorTest, InvalidIdentitiesAreNotExtracted)
...
+      "current-span",
+      nostd::shared_ptr<trace::Span>(new trace::DefaultSpan(trace::SpanContext::GetInvalid()))};
+  context::Context ctx2 = propagator->Extract(carrier, ctx1);
+
+  TextMapCarrierTest carrier2;
+  propagator->Inject(carrier2, ctx2);
+
+  EXPECT_TRUE(carrier.headers_.count("traceparent") > 0);
+  EXPECT_TRUE(carrier.headers_.count("tracestate") > 0);
+  EXPECT_EQ(carrier.headers_["tracestate"], "congo=t61rcWkgMzE");
```
Comment (ThomsonTan) [inline api/test/trace/propagation/http_text_format_test.cc:196]: Assigned `"congo=t61rcWkgMzE"` a name as it is used twice?

File: api/include/opentelemetry/trace/propagation/text_map_propagator.h (hunk 4 above)
Comment (lalitb) [reply api/include/opentelemetry/trace/propagation/text_map_propagator.h]: Thanks. fixed.

Comment (lalitb) [reply api/include/opentelemetry/trace/propagation/text_map_propagator.h]: Thanks. fixed.

**Review by ThomsonTan — APPROVED** [review]

**Review by jsuereth — APPROVED** [review]

**Pull request merged**